src/main/java/carpet/utils/Messenger2.java

```diff
@@ -0,0 +1,247 @@
+        DARK_BLUE   ('v', (s, f) -> s.withColor(ChatFormatting.DARK_BLUE)), // navy
+        BLACK       ('k', (s, f) -> s.withColor(ChatFormatting.BLACK)),
+
+        COLOR       ('#', (s, f) -> {
+            TextColor color = TextColor.parseColor("#"+f);
+            return color == null ? s : s.withColor(color);
+        }, s -> {
+            Matcher m = colorExtract.matcher(s);
+            return m.find() ? m.group(1) : null;
+        }),
+        ;
+
+        public char code;
+        public BiFunction<Style, String, Style> applier;
+        public Function<String, String> container;
+        CarpetFormatting(char code, BiFunction<Style, String, Style> applier)
+        {
+            this(code, applier, s -> s.indexOf(code)>=0?Character.toString(code):null);
+        }
+        CarpetFormatting(char code, BiFunction<Style, String, Style> applier, Function<String, String> container)
+        {
+            this.code = code;
+            this.applier = applier;
+            this.container = container;
+        }
+        public Style apply(String format, Style previous)
+        {
+            String fmt;
+            if ((fmt = container.apply(format))!= null) return applier.apply(previous, fmt);
+            return previous;
+        }
+    };
+
+    Messenger2(Object... data) {
+        for (Object x : data) {
+            this.data.add(x);
+        }
+        this.size = this.data.size();
+        c = 0;
+    }
+
+    /*
+     * composes single line, multicomponent message, and returns as one chat
+     * messagge
+     */
+    public Component getSomeComponents(){
+        return getSomeComponents("|");
+    }
+    public Component getSomeComponents(String str) {
+        TextComponent message = new TextComponent("");
+        while (true) {
+            if (isEnd(str)) {
+                return message;
+            }
+            Component com = getAComponent();
+            message.append(com);
+        }
+    }
+
+    private Component getAComponent() {
+        BaseComponent base = (BaseComponent) getARawComponent();
+        while (isModifier(now())) {
+            base = (BaseComponent) getModifier(((String) now()).charAt(0)).apply(base, (String) now());
+        }
+        return base;
+    }
+
+    private BiFunction<BaseComponent, String, Component> getModifier(char ch) {
+        c += 1;
+        return (base, message) -> {
+            Style previousStyle = base.getStyle();
+            base.setStyle(switch (ch) {
+                case '?' -> previousStyle
+                        .withClickEvent(new ClickEvent(ClickEvent.Action.SUGGEST_COMMAND, message.substring(1)));
+                case '!' -> previousStyle
+                        .withClickEvent(new ClickEvent(ClickEvent.Action.RUN_COMMAND, message.substring(1)));
+                case '^' -> previousStyle
+                        .withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, getSomeComponents()));
```

Comment:
maybe i should use getAComponent() to replace getSomeComponents()
dont know which is better.